Disable back button when no subscriptions are available; See: websharks/comment-mail#229

kristineds opened the pull request:
Disable back button when no subscriptions are available; 

- Continuation from this Pull Request: https://github.com/websharks/comment-mail-pro/pull/71
- Tested in the following scenarios:
    - User is not logged in and no Subscription Key is present
    - User is logged in but does not have any subscriptions

See: websharks/comment-mail#229

--------------------------------

**TODO**
- [ ] <del> Display back button and "Manage Subscriptions" link when user is logged in with existing subscriptions. </del>

kristineds commented:
@raamdev I started over from scratch and opened a new Pull Request, like you suggested. When I attempted to reproduce the problem (using the copy from dev branch), it was working as before, i.e. user can view the Back button and "Manage Subscriptions" link and it redirects to the correct page. There was no error message. 

With this current PR, after testing it with users who **do have** subscriptions, the Back button and "Manage Subscriptions" link are not being displayed at all. I've been racking my brain, but I'm coming up empty. Could you please point me in the right direction?

Note that I used the lines that @jaswsinc recommended here: https://github.com/websharks/comment-mail/issues/229#issuecomment-202682207

raamdev commented on src/includes/templates/type-a/site/footer-tag.php:
```diff
@@ -31,7 +31,9 @@
 
 // Summary return URL; w/ all summary navigation vars preserved.
-$sub_summary_return_url = $plugin->utils_url->subManageSummaryUrl(!empty($sub_key) ? $sub_key : '', null, true);
+$current_email     = (boolean)$this->plugin->utils_sub->currentEmail();
```
There's a bug on this line: Casting this as `(boolean)` means that when you try to pass `$current_email` to `subManageSummaryUrl()` on the following line, a boolean value is getting passed instead of the actual email address (which is what `subManageSummaryUrl()` expects). This should not be cast as boolean.

kristineds replied:
@raamdev I have casted that line as a `(boolean)` since you previously suggested that it should be added.  https://github.com/websharks/comment-mail-pro/pull/71#discussion_r67524902.

raamdev commented on src/includes/templates/type-a/site/footer-tag.php:
```diff
@@ -31,7 +31,9 @@
 // Summary return URL; w/ all summary navigation vars preserved.
-$sub_summary_return_url = $plugin->utils_url->subManageSummaryUrl(!empty($sub_key) ? $sub_key : '', null, true);
+$current_email     = (boolean)$this->plugin->utils_sub->currentEmail();
+$has_subscriptions = $current_email ? (boolean)$this->plugin->utils_sub->queryTotal(null, ['sub_email' => $current_email, 'status' => 'subscribed', 'sub_email_or_user_ids' => true]) : false;
```
Here you can cast `$current_email` as boolean, since you're doing a boolean comparison, e.g., `$has_subscriptions = (boolean)$current_email ?...`.

raamdev commented:
@jaswsinc Is there a reason that [`currentEmail()`](https://github.com/websharks/comment-mail-pro/blob/160618/src/includes/classes/UtilsSub.php#L1186) requires `current_user_can('edit_posts')`?? If I'm logged in as a Subscriber and I have subscriptions, and my Comment Mail settings have `all_wp_users_confirm_email` set to false, then that `currentEmail()` function returns an empty value (i.e., not possible) instead of returning the email address of the account I'm logged in with. That seems wrong to me (and is probably part of the bug Kristine is seeing here).

jaswsinc commented:
@raamdev There is, yes. `currentEmail()` won't return the email address you're logged in with whenever `all_wp_users_confirm_email` is `false`, because users who don't confirm their email address when registering are able to register with any email address they choose; i.e., perhaps a fake address. This would allow, for example, a user to register on the site with a fake email address and then manage subscriptions that belong to the person who _actually_ owns the email address associated with a CM 'Subscriber'.

So while CM _does_ try to make a connection between 'Subscribers' and 'Users' of the site whenever there is a secure way to accomplish that, we _do not_ rely upon WP users for any sort of authentication. Instead, [this cookie](https://github.com/websharks/comment-mail-pro/blob/160618/src/includes/classes/UtilsSub.php#L1191) is returned by that same routine. 

That cookie is how Comment Mail identifies the owner of an email address via email confirmation. As soon as you've confirmed at least one subscription, you will get that cookie and then `currentEmail()` will be capable of returning the current 'Subscriber' email address, no matter which account you are logged into WordPress with.

The exceptions to this, being:

- `all_wp_users_confirm_email` is `true` instead of being the default value `false`.
- Or, you're logged into an account that has a higher trust level; e.g., you can `edit_posts`.

In those cases, we consider the current 'User' to be the current 'Subscriber', giving the current WP User precedence over the CM-specific cookie.

kristineds commented:
@raamdev @jaswsinc I've updated the code and moved the `boolean` for the `$current_email` but I'm still encountering the same issue after testing, i.e. A user logged in with subscription can't view the Back button and "Manage Subscriptions" link.

raamdev commented:
@jaswsinc writes...
> users who don't confirm their email address when registering are able to register with any email address they choose; i.e., perhaps a fake address. This would allow, for example, a user to register on the site with a fake email address and then manage subscriptions that belong to the person who actually owns the email address associated with a CM 'Subscriber'.

Ah, that makes sense. Thank you.

@jaswsinc writes...
> As soon as you've confirmed at least one subscription, you will get that cookie and then currentEmail() will be capable of returning the current 'Subscriber' email address, no matter which account you are logged into WordPress with.

Hmm, so that means there's no good way to access the 'My Comment Subscriptions' list without going through a link in one of the emails to ensure that the cookie is set? 

The issue that we're trying to solve is what Kristine mentioned in her last comment above. ↑ How do logged-in users (with a Subscriber role) view a list of their subscriptions when `all_wp_users_confirm_email` is `false`?

jaswsinc commented:
> Hmm, so that means there's no good way to access the 'My Comment Subscriptions' list without going through a link in one of the emails to ensure that the cookie is set?

That is correct, yes. Once you have confirmed at least one subscription, your browser will have that cookie on future visits. So once you actually become a 'Subscriber', you can see your list of subscriptions, but not before. You won't have any subscriptions until you confirm anyway.

> The issue that we're trying to solve is what Kristine mentioned in her last comment above. ↑ How do logged-in users (with a Subscriber role) view a list of their subscriptions when all_wp_users_confirm_email is false?

If `all_wp_users_confirm_email` is `false`, the only way to see a list of your subscriptions is if you have the Comment Mail cookie; i.e., you will need to have confirmed at least one subscription and still have the cookie set in your browser.

Or, you will need to click a link from one of the emails that you receive, which contains a key which authenticates you, and results in that cookie being set again and again automatically as you come back to the site. For example, you can review the calls to `setCurrentEmail()` in action handlers [like these](https://github.com/websharks/comment-mail-pro/blob/160618/src/includes/classes/SubManageActions.php#L88).

raamdev commented on src/includes/classes/SubManageSubFormBase.php:
```diff
@@ -159,6 +159,8 @@ protected function maybeDisplay()
         $sub_key = $this->sub_key;
         $is_edit = $this->is_edit;
         $sub     = $this->sub;
+        $current_email     = (boolean)$this->plugin->utils_sub->currentEmail();
```
@kristineds The same casting-as-boolean bug is present on this line too. `$current_email` is being forced to a boolean value, so the next line that passes `$current_email` to the `queryTotal()` method is not getting the data it's expecting (i.e., an email address).

raamdev commented:
@jaswsinc writes...
> If all_wp_users_confirm_email is false, the only way to see a list of your subscriptions is if you have the Comment Mail cookie

Got it. We'll need to improve the UX for this as it's really, really confusing right now. I understand and agree with the reasoning behind the way it works, i.e., that if a user doesn't need to confirm their email address when registering for an account, they could just create an account using someone else's email address and then view/edit all of their subscriptions — not good. 

I also see that most of this is already explained within the plugin, but I did not read any of this in detail until just now (mostly because it's a lot to read, but also because it didn't seem relevant to my problem: I'm trying to understand why my users can't see a list of their subscriptions, not something about Auto-Confirm Settings). 

![2016-08-10_14-50-13](https://cloud.githubusercontent.com/assets/53005/17566634/36b5345c-5f0a-11e6-85af-c90cd478dc6c.png)

To improve the UX here, I suggest that we have a way for users to request a link to view their subscriptions, a form where they can input their email address and receive a link to view a list of their subscriptions. I'll open a separate GitHub issue for that.

raamdev commented:
@kristineds All this PR needs to ensure is the following:

- The "My Comment Subscriptions" link and the Back Button should always appear on the Edit Subscription page, which right now can only be reached by clicking the "My Comment Subscriptions" link in a comment notification email, or by clicking "manage my subscriptions" on the comment form when you've already been authenticated (see below for how this works).

     ![2016-08-10_14-56-15](https://cloud.githubusercontent.com/assets/53005/17566962/6c443d56-5f0b-11e6-833a-94852693b834.png)

- When **Comment Mail → Config. Options → Auto-Confirm Settings → All WordPress Users Confirm their Email Address?** is set to **Yes** and a user who has comment subscriptions is logged in, they should see a link on the comment form that says "Manage My Subscriptions". Clicking that link should allow them to view and edit their subscriptions. 

     ![2016-08-10_15-03-17](https://cloud.githubusercontent.com/assets/53005/17567076/e2f0d63a-5f0b-11e6-8b25-7b6609345e0c.png)

- When **Comment Mail → Config. Options → Auto-Confirm Settings → All WordPress Users Confirm their Email Address?** is set to **No** (the default) and a user who has comment subscriptions is logged in, they should NOT see a link on the comment form that says "Manage My Subscriptions" UNLESS they have previously clicked the "My Comment Subscriptions" link in a comment notification email, which creates a cookie in their browser containing the key that's needed to authenticate that they should actually see a list of their subscriptions.

jaswsinc commented:
> To improve the UX here, I suggest that we have a way for users to request a link to view their subscriptions, a form where they can input their email address and receive a link to view a list of their subscriptions. I'll open a separate GitHub issue for that.

Sounds great!

kristineds commented:
> The "My Comment Subscriptions" link and the Back Button should always appear on the Edit Subscription page,

@raamdev ⬆️  Just tested this and it doesn't work in my recent PR. https://github.com/websharks/comment-mail-pro/pull/82/commits/64f1aeafbd13e13a765b01172e0121f79f3c997d

> When Comment Mail → Config. Options → Auto-Confirm Settings → All WordPress Users Confirm their Email Address? is set to Yes and a user who has comment subscriptions is logged in, they should see a link on the comment form that says "Manage My Subscriptions". 

⬆️  Same here. The Back button and  "Manage My Subscriptions" link is missing. 

> When Comment Mail → Config. Options → Auto-Confirm Settings → All WordPress Users Confirm their Email Address? is set to No (the default) and a user who has comment subscriptions is logged in, they should NOT see a link on the comment form that says "Manage My Subscriptions" UNLESS they have previously clicked the "My Comment Subscriptions" link in a comment notification email, which creates a cookie in their browser containing the key that's needed to authenticate that they should actually see a list of their subscriptions.

⬆️  This works but the "My Comment Subscriptions" link is pointing to the wrong URL.

raamdev commented on src/includes/templates/type-a/site/footer-tag.php:
```diff
@@ -31,7 +31,9 @@
-$sub_summary_return_url = $plugin->utils_url->subManageSummaryUrl(!empty($sub_key) ? $sub_key : '', null, true);
+$current_email     = $this->plugin->utils_sub->currentEmail();
+$has_subscriptions = (boolean)$current_email ? (boolean)$this->plugin->utils_sub->queryTotal(null, ['sub_email' => $current_email, 'status' => 'subscribed', 'sub_email_or_user_ids' => true]) : false;
+$sub_summary_return_url = $has_subscriptions && $plugin->utils_url->subManageSummaryUrl(!empty($sub_key) ? $sub_key : '', null, true);
```
@kristineds I suggest playing around with this line. I think the issue you mentioned [here](https://github.com/websharks/comment-mail-pro/pull/82#issuecomment-238984950) is related to this line.

kristineds replied:
@raamdev Should it be like this?

```
$sub_summary_return_url = (boolean)$has_subscriptions ? $plugin->utils_url->subManageSummaryUrl($sub_key, null, true) : false;
```

However, I tested this but it's not working. 😞  Any more clue on how to go about this?

raamdev commented:
@kristineds I fixed this in https://github.com/websharks/comment-mail-pro/pull/82/commits/724946ffb49130376d3bb3892298c28b8c4f0bdb. This PR is now ready to merge. I'll merge it shortly.

raamdev commented:
@kristineds Merged! Thank you. :-)
